---
title: "Tyrosine recombinase subfamily annotation pipeline"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```


## Research Objective 1: 

**Annotate tyrosine integrase proteins from MGE databases to subfamilies from Smyshlaev et al. (2021) using SMART reference seqeunces**

Smyshlaev et al's HMMs are available in their [supplementary data](https://www.embopress.org/action/downloadSupplement?doi=10.15252%2Fmsb.20209880&file=msb20209880-sup-0001-Appendix.pdf). Their HMMs don't include all the available sequences from the SMART nr database though. We could possibly improve the HMM somewhat by including the extra data.


### SMART reference data

Sequences for each domain subfamily can be downloaded from:
http://smart.embl-heidelberg.de/smart/search.cgi?keywords=tyrosine+recombinase. It is a bit tricky to find the link for each sub-family, and from what I can tell, there is no static links for programmatic downloads.

To download the sequences, go to each domain subfamily page (eg. Cyan), click the number of proteins in the domain database: eg. 3164 in "There are 3164 Cyan domains in 3164 proteins in SMART's nrdb database.". On the 'All proteins containing <name> domain' page, change 'Action' to 'download protein sequences as a FASTA file'. In the next box, select 'specific domain only' download the text to `./data/SMART/domains_fasta/`. (We can choose full sequences or specific domain here. Download the full proteins are in `./data/SMART/full_protein_fasta/`.


The data have been collected in R as follows:


```{r SMART fasta}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("Biostrings")
library(Biostrings)
library(tidyverse)

# make dataframe of fastas for each subfamily of domains
subfamilies <- 
  tibble(path = Sys.glob('./data/SMART/domain_fasta/*.fasta')) |> 
  mutate(name = str_remove_all(path, './data/SMART/domain_fasta/|\\.fasta'),
         fasta = map(path, readAAStringSet),
         nseq = map_int(fasta, ~length(.x)),
         domain_names = map(fasta, ~names(.x)),
         domain_seqs = map(fasta, paste)
         ) |> 
  left_join(read_csv('./data/SMART/SMART_YR_ids.csv') |> 
              janitor::clean_names()
            )  |>  
  select(name, id, nseq, everything()) |> 
  arrange(nseq)

glimpse(subfamilies)
```


### Multiple sequence alignment with DECIPHER

I aligned the amino acid sequences for each subfamily with the decipher package with default settings, as follows:

```{r alignment, eval=FALSE, include=TRUE}
# BiocManager::install("DECIPHER")
library(DECIPHER)

# Create alignment using decipher from fasta file, given path
make_alignment <- function(name, path){
  message(paste("\n\nAligning subfamily:", name))
  aligned <- AlignSeqs(readAAStringSet(path))
  outpath <- paste0('./data/SMART/domain_alignments/', name, '.aln')
  writeXStringSet(aligned, filepath = outpath)
  return(aligned)
}

# Do all the alignments. Will create .aln files
subfamilies <- subfamilies |> 
  mutate(aligned = map2(.x = name, .y = path, .f = ~make_alignment(.x,.y)))
```

This creates the aligned fasta files in `./data/SMART/domain_alignments' as they are finished. The Int_SXT, Int_Tn916, and Xer sub-families take many hours to align.



### Building hidden Markov models

I installed HMMER through conda and created HMMs for each 'phage integrase' domain sub-family.

```
conda install -c bioconda hmmer

# cd into the `./data/SMART/` folder
mkdir ./domain_hmm

# loop to build model for each subfamily
for file in ./domain_alignments/*.aln
do
  outfile=`echo $file | sed 's/alignments/hmm/g;s/\\.aln/.hmm/g'`
  hmmbuild $outfile $file
  echo $outfile
done

# compile the models into a single hmm database
cat domain_hmm/* > subfamilies.hmm

# compress and index the hmm files
hmmpress subfamilies.hmm

```

I now have a database of domain models that I can use to annotate the SMART domains in any query protein.

The statistics for each subfamily were recorded from the output of `hmmbuild` 
```{r hmm graphs}
hmms_output <- read_csv('./data/SMART/hmm_output.csv')
glimpse(hmms_output)
# nseqs
a <- hmms_output |> 
  ggplot(aes(nseq, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n sequences',
       subtitle = 'Number of sequences for profile HMMs by sub-family')
# alignment columns
b <- hmms_output |>
  ggplot(aes(alen, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n columns',
       subtitle = 'Number of alignment columns by sub-family')
# number of consensus positions in hmm
c <- hmms_output |>
  ggplot(aes(mlen, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n nodes',
       subtitle = 'Number of positions in HMM by subfamily')
# effective number of sequences
d <- hmms_output |>
  ggplot(aes(eff_nseq, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n sequences',
       subtitle = 'Effective number of sequences by sub-family')
```


```{r, fig.width=8}
library(patchwork)
((a+b)|(c+d)) + plot_layout(nrow = 4)
```
 

 <!-- Hmmbuild output shows some data that might be of interest-->
 <!-- - n_eff: effective total seq number -->
 <!-- - re/pos: relative entropy per position -->
 <!-- .... -->
 
 
## Check annotations of full proteins from SMART subfamilies (hmmscan)

I used hmmscan with the database of SMART hmms to align & score domains in the full sequences of all the SMART proteins. I think that the scores should indicate the discriminatory power of the ensemble of models for differentiating the subfamilies; i.e. if scores are close for multiple domains for a given sequence then it will be difficult to classify domains accurately and vice-versa if the scores are very different across hmms.
```
# in the directory with all the SMART data
cd ./data/SMART

# scan all smart integrases using all 20 subfamily hmms
hmmscan -o smart.scan.raw --tblout smart.scan.tbl --domtblout hmmscan -o smart.scan.raw --tblout smart.scan.tbl --domtblout smart.scan.domtbl --pfamtblout smart.scan.pftblout subfamilies.hmm SMART_full_proteins.fasta 

# store output in hmmscan_results
mkdir hmmscan_results
mv smart.scan.* hmmscan_results
```

Save the output to file; still need to figure out how to parse and link with true domains.

<!-- ## Search the sequence database with hmmsearch -->

 <!-- hmmsearch accepts any fasta, embl/uniprot text, and genbank format files, with automatic detection. -->

<!-- hmmsearch takes a single hmm and sequences and scores hits in the sequences. -->

<!-- ``` -->
<!-- hmmsearch --tblout <outfile> < family.hmm > < sequences.fa >   > < outfile > -->
<!-- ``` -->

<!-- Parse hmmsearch results into R -->

<!-- ```{r} -->
<!-- library(rhmmer) -->
<!-- ritc_ritc <- rhmmer::read_tblout(file = './data/SMART/hmmsearch_ritc_ritc.out') -->
<!-- glimpse(ritc_ritc) -->
<!-- ``` -->


```{r}
# need to create a database of SMART proteins to establish the gathering threshold and to test the pipeline.

```
 

 
## Pfam HMMs for annotating other domains

I downloaded the Pfam-A.seed alignment from ftp.ebi.ac.uk.  
I built the Pfam HMMs database with HMMER.

```
# unpack seed alignments
tar -xvf Pfam-A.seed.gz

# build hmm for each alignment
hmmbuild Pfam-A.hmm Pfam-A.seed

# compress and index the flatfile; 
# create 4 binary files .h3m .h3i .h3f and .h3p
hmmpress Pfam-A.hmm


```


```{r}
pfam_clans <- read_tsv('./data/Pfam-A/Pfam-A.clans.tsv',
         col_names = c('family_id','clan', 'family_name', 'description'))

``` 
 
 
 
 
## CDD batch search

**Input fasta ids (> lines) have to not have anything after a space. And the pipe symbol messes it up too.**  
Batch can handle 4000 queries.

CDD batch search with SMART does not detect the same integrase subfamilies - seems like it hasn't been updated since these new domains were added.

CDD domains match SMART domains??

*SMART has a batch annotation tool, though: 'You will only get results for sequences that have a match in SMART database, ie. only precomputed results will be displayed'.*


CDD has web api or can use stand-alone RPS-blast with CDD data from ftp.ncbi.nih.gov > cdd.  

HTTP GET command to send request
```{r, eval=FALSE}
library(httr)
base <- "https://www.ncbi.nlm.nih.gov/Structure/bwrpsb/bwrpsb.cgi?"

# databases: "cdd," "pfam," "smart," "tigrfam," "cog," "kog"
db <- 'db=CDD' 

# tdata: hits, aligns, or feats
tdata <- 'tdata=hits'

# data output: rep, std, or full
dmode <- 'dmode=std'

# sequences
# queries <- input_fasta
```
 
 
 








