---
title: "Tyrosine recombinase subfamily annotation pipeline"
output:
  html_notebook:
    theme: paper
    toc: yes
    toc_float: 
        collapsed: false
        smooth_scroll: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```


```{r libraries, include=FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("Biostrings")
library(Biostrings)
library(DECIPHER)
library(tidyverse)
library(janitor)
library(furrr)
library(rentrez)
```
   
### Objective:    
    
**Annotate tyrosine integrase proteins from MGE databases to subfamilies from Smyshlaev et al. (2021) using SMART reference sequences.**    


----

### SMART reference data

     
Smyshlaev et al's HMMs are available in their [supplementary data](https://www.embopress.org/action/downloadSupplement?doi=10.15252%2Fmsb.20209880&file=msb20209880-sup-0001-Appendix.pdf). Their HMMs don't include all the available sequences from the SMART nr database though. We could possibly improve the HMM somewhat by including the extra data available in the SMART database.   
  
Sequences for each domain subfamily can be downloaded from:
http://smart.embl-heidelberg.de/smart/search.cgi?keywords=tyrosine+recombinase. It is a bit tricky to find the link for each sub-family, and from what I can tell, there are no static links or API for programmatic downloads.

To download the sequences, go to each integrase catalytic domain subfamily page (eg. Cyan) after searching for 'tyrosine recombinase', click the number of proteins in the domain database (eg. the 3164 in "There are 3164 Cyan domains in 3164 proteins in SMART's nrdb database."). On the 'All proteins containing <name> domain' page, change 'Action' to 'download protein sequences as a FASTA file'. In the next box, select 'specific domain only' download the text to a file in `./data/SMART/domains_fasta/`. Then go back and choose full sequences and download the full proteins to `./data/SMART/full_protein_fasta/`.  
  
I collected all the domain sequences from these SMART integrase subfamilies fasta files into an R dataframe (`ref_domains`) and did the same for the full proteins (`ref_integrases`). Then, I joined the domains and full sequence files into a single dataset for all the integrases from SMART (`SMART_db`). The total number of domains and proteins for each family do not necessarily match because a couple have multiple domains

(code from script: `./code/tidy_smart_data.R`)

```{r SMART domains, message=FALSE, warning=FALSE}
# make dataframe of fastas for each subfamily of domains
ref_domains <- 
  tibble(dom_path = Sys.glob('./data/SMART/domain_fasta/*.fasta')) |>
  mutate(
    subfamily = str_remove_all(dom_path, './data/SMART/domain_fasta/|\\.fasta'),
    dom_fasta = map(dom_path, readAAStringSet),
    dom_n = map_int(dom_fasta, ~length(.x)),
    dom_name = map(dom_fasta, ~names(.x)),
    dom_seq = map(dom_fasta, paste)
  ) |> 
  # add accession numbers
  left_join(read_csv('./data/SMART/smart_subfamily_ids.csv',
                     col_types = cols()) |> 
              janitor::clean_names() |> 
              select(-n_domains),
            by = 'subfamily') |> 
  select(subfamily, id, everything())

# All the full protein sequences for 20 subfamilies from SMART
ref_integrases <- 
  tibble(
    prot_path = Sys.glob('./data/SMART/full_protein_fasta/*.fasta')) |> 
  mutate(
    subfamily = str_remove_all(
      prot_path, './data/SMART/full_protein_fasta/|\\_proteins.fasta'),
    prot_fasta = map(prot_path, readAAStringSet),
    prot_n = map_int(prot_fasta, ~length(.x)),
    prot_name = map(prot_fasta, ~names(.x)),
    prot_seq = map(prot_fasta, paste)
  ) |> 
  # add accession numbers
  left_join(read_csv('./data/SMART/smart_subfamily_ids.csv', 
                     col_types = cols()) |> 
              janitor::clean_names() |> 
              select(-n_domains),
            by = 'subfamily') |> 
  select(subfamily, id, everything())

# join the full proteins and domains data together by subfamily
SMART_db <- 
  left_join(ref_domains, ref_integrases, by = c('subfamily', 'id')) |> 
  mutate(subfamily = as_factor(subfamily))

glimpse(SMART_db)
write_rds(SMART_db, './data/SMART_db.rds', compress = 'gz')

rm(ref_domains, ref_integrases, SMART_db)
```  
\  
  
----

### Non-integrase sequences (negative examples)

scripts:
- `./code/get_refseq_non_integrases.R`
- `./code/tidy_non_integrases.R`

**Specific proteins from RefSeq (by name)**

I obtained some sequences (~23k) that are (almost certainly) not tyrosine recombinases by searching RefSeq for specific protein names. Here, I mainly focus on proteins with similar DNA-binding and nuclease activity. 

**Eukaryotic proteomes from Uniprot (non-integrases)**

Human, yeast and arabidopsis reference proteomes will be used as representative eukaryote sequences to classify as non-integrases. Reference proteomes have one sequence per protein.

**Pfam domain families**

Several DDE transposases, HTH transposases, .... *could add a variety of others, where to stop though?*

```{r}
read_rds('./data/non_integrases_df.rds)
```



### Multiple sequence alignment of subfamilies

I aligned the domain AA sequences for each subfamily with the decipher package (with default settings). This creates the aligned fasta files in `./data/SMART/domain_alignments` as they are finished. The Int_SXT, Int_Tn916, and Xer sub-families take many hours to align.

<!-- # ```{r MSA, eval=FALSE, include=TRUE} -->
```
# BiocManager::install("DECIPHER")
library(DECIPHER)

# Create alignment of for each subfamily using decipher; 
# reads fasta files from glob of paths
# writes alignments as they finish
make_alignment <- function(name, path){
  message(paste("\n\nAligning subfamily:", name))
  aligned <- AlignSeqs(readAAStringSet(path))
  # outpath <- paste0('./data/SMART/domain_alignments/', name, '.aln')
  # writeXStringSet(aligned, filepath = outpath)
  return(aligned)
}

# Do all the alignments. 
# Creates .aln files in ./data/SMART/domain_alignments 
# Don't need to save obj when finished
# Takes 1 day or more
domains <- read_rds('./data/SMART_db.rds')
domains <- domains |> 
  mutate(aligned = map2(.x = subfamily, 
                        .y = dom_path,
                        .f = ~make_alignment(.x,.y)))
```


----

### Building hidden Markov models

I installed HMMER (v3.3.2) through Conda and created HMMs for each 'phage integrase' domain sub-family using the multiple alignments. I concatenated have a database of domain models that I can use to annotate the SMART domains in any query protein.


<!-- ```{bash hmmbuild, eval = FALSE, include = TRUE} -->
```
# in bash
conda install -c bioconda hmmer
cd ./data/SMART/
mkdir ./domain_hmm

# loop to build model for each subfamily
for file in ./domain_alignments/*.aln
do
  outfile=`echo $file | sed 's/alignments/hmm/g;s/\\.aln/.hmm/g'`
  echo $outfile
  hmmbuild $outfile $file
done

# **might not need this bit if not using hmmscan**
# compile the models into a single hmm database for hmmscan
# compress and index the hmm files
cat domain_hmm/* > subfamilies.hmm
hmmpress subfamilies.hmm
```
\  
\  


### Search for domains in protein sequences with hmmsearch

I did an `hmmsearch` for each subfamily HMM against the collection of SMART integrases. `hmmsearch` takes a single HMM and a set of protein sequences, returning scores for all significant hits in the sequences. Significant hits are found for multiple subfamilies though, so we'll still need a precise way to discriminate between classes.

<!-- ```{bash hmmsearch, eval = FALSE, include = TRUE} -->
```
# bash
mkdir hmmsearch_res

for hmm in domain_hmm/*;
  do
  outfile=`echo $hmm | sed 's|domain_hmm/||g;s|.hmm|.search|g'`
  echo $outfile
  hmmsearch -o hmmsearch_res/$outfile.raw --tblout hmmsearch_res/$outfile.tbl --domtblout hmmsearch_res/$outfile.domtbl $hmm all_SMART_integrases.fa
done
```
\  

I did the same HMM alignment scoring for non-integrase sequences
``` 
# bash

```
  
  
I parsed all the `hmmsearch` results tables and rectangled the data by joining the bitscores to the protein names with their true subfamilies in the `SMART_db` dataframe.  
  
```{r join hmmsearches}
# read hmmsearch tblout result
# keep only the best_domain_score column, 
# rename it as the subfamily name for the hmm
read_hmmersearch_tbl <- function(path){
  read_table(
    file = path,
    na = '-',
    comment = '#',
    col_types = cols(),
    col_names =  c(
      "prot_name", "prot_accession", 
      "hmm_name", "hmm_accesion",
      "seq_eval", "seq_score", "seq_bias",
      "best_dom_eval", "best_dom_score", "best_dom_bias",
      'domain_n_exp', 'domain_n_reg', 'domain_n_clu', 'domain_n_ov',
      'domain_n_env', 'domain_n_dom', 'domain_n_rep', 'domain_n_inc',
      "description")
  ) |> 
    select(prot_name, hmm_name, best_dom_score) |> 
    distinct() |> 
    pivot_wider(names_from = hmm_name, values_from = best_dom_score) |> 
    unnest(cols = c())
}

# All the full protein sequences for 20 subfamilies from SMART
# Unnest and split protein names to match hmmsearch results
integrases <- 
  read_rds('./data/SMART_db.rds') |> 
  select(subfamily, id, prot_name, prot_seq) |> 
  unnest(cols = c(prot_name, prot_seq)) |> 
  mutate(temp = trimws(str_extract(prot_name, '.*? ')),
         prot_description = str_remove(prot_name,  '.*?\\s'),
         prot_name = temp) |> 
  select(-temp)

# read all the results and join into one table
# join to ref_integrases table by protein names, such that each row has the
# scores across all models for a given sequence
hmmsearches <- 
  Sys.glob("./data/SMART/hmmsearch_res/*.search.tbl") |> 
  map(read_hmmersearch_tbl) |> 
  purrr::reduce(left_join, by = 'prot_name') |> 
  right_join(integrases, by = 'prot_name') |> 
  select(prot_name, prot_description, subfamily, id, everything())

glimpse(hmmsearches)
rm(integrases, read_hmmersearch_tbl)

write_rds(hmmsearches, './data/smart_refseqs_hmm_scores.rds', compress = 'gz')
```
\  
\  
I plotted the score profiles for each subfamily to explore the discriminative power of the bit-scores. This shows the distribution of scores for a set of sequences from a given subfamily (panel) against each subfamily hmm (y-axis), in terms of the bit scores (x-axis). Mostly the true subfamily has the best-scoring HMM alignment but in the case of `Int_Tn916` and `Int_BPP-1` integrases, the scores are very similar between the two HMMs.
   
```{r plot bitscores}
hmmsearches |> 
  pivot_longer(cols = c(Arch1:Xer), 
               names_to = 'hmm_dom', values_to = 'hmm_score') |> 
  mutate(hmm_score = replace_na(hmm_score, 0)) |> 
  mutate(highlight = ifelse(hmm_dom == subfamily, T, F)) |> 
  ggplot(aes(x = hmm_score, y = fct_rev(hmm_dom),
             color = highlight)) +
  geom_violin() +
  labs(x = 'Best domain bit score', y = 'Hidden Markov model', 
       subtitle = 'HMM bit score profiles for each set of tyrosine integrases from the SMART subfamilies') +
  facet_wrap(~subfamily, nrow = 3) +
  theme_light() + 
  theme(legend.position = 'null', 
        axis.text.y = element_text(size = 5),
        panel.grid.major.x = element_blank()
        )

```
 
We could also exchange the HMMs and true subfamilies in the plot above to get the scores for each model as a panel, with the true classes on the y-axis.

```{r plot bitscores2}
hmmsearches |> 
  pivot_longer(cols = c(Arch1:Xer), 
               names_to = 'hmm_dom', values_to = 'hmm_score') |> 
  mutate(hmm_score = replace_na(hmm_score, 0)) |> 
  mutate(highlight = ifelse(hmm_dom == subfamily, T, F)) |> 
  ggplot(aes(x = hmm_score, y = fct_rev(subfamily),
             color = highlight)) +
  geom_violin() +
  # ggthemes::geom_tufteboxplot(stat = "fivenumber",
  #                             size = 1.5, whisker.type = 'line') +
  facet_wrap(~hmm_dom, nrow = 2) +
  labs(x = 'Best domain bit score', y = 'True subfamily', 
       subtitle = 'HMM bit score profiles for each hidden Markov model') +
  theme_light() + 
  theme(legend.position = 'null', 
        axis.text.y = element_text(size = 5))
```



```{r}
rm(hmmsearches)
```




### k-mer profile features

```{r}
library(Biostrings)

# function generates sequence feature columns [(1-4)-mers] for classification
generate_kmer_features <- function(df, k){
  df <- as.data.frame(df)
  df$seq <- Biostrings::DNAStringSet(df$seq)
  features.df <- df %>%
    cbind(
      Biostrings::oligonucleotideFrequency(df$seq, k, as.prob = TRUE)
    ) %>%
    dplyr::select(-seq) 
  return(features.df)
}
```


### Train a classifier

#### Split data / Resampling
 
```{r split data, message=FALSE, warning=FALSE}
library(tidyverse)
library(tidymodels)

set.seed(1)

# make a small dataset with stratified sampling for code development
df <- 
  read_rds('./data/smart_refseqs_hmm_scores.rds') |>
  group_by(subfamily) |> 
  sample_frac(0.1)

summary(df$subfamily)

# split train and test, stratified by true subfamily
df_split <- initial_split(df, strata = subfamily)
df_train <- training(df_split)
df_test <- testing(df_split)

# set up resampling scheme
cv <- nested_cv(df_train, v = 5, strata = subfamily)

```


```{r}

# create model specifications

# set up hyperparameter search ranges

# create data preprocessing specification

# do parameter tuning through CV

# examine performance metrics

# explore stacking / ensembles

```
 
 
 
### Pfam HMMs for annotating other domains

It might be useful to add other domains found in tyrosine recombinases, specifically the arm-binding domain and the SAM domains

I downloaded the Pfam-A.seed alignment from ftp.ebi.ac.uk.  
I built the Pfam HMMs database with HMMER.

```
# unpack seed alignments
tar -xvf Pfam-A.seed.gz

# build hmm for each alignment
hmmbuild Pfam-A.hmm Pfam-A.seed

# compress and index the flatfile; 
# create 4 binary files .h3m .h3i .h3f and .h3p
hmmpress Pfam-A.hmm
```


Other phage integrase domains that are worth annotating / using for classification
    
    #=GF ID   Arm-DNA-bind_1
    #=GF ID   Arm-DNA-bind_2
    #=GF ID   Arm-DNA-bind_3
    #=GF ID   Arm-DNA-bind_4
    #=GF ID   Arm-DNA-bind_5
    #=GF ID   Phage_int_SAM_1
    #=GF ID   Phage_int_SAM_2
    #=GF ID   Phage_int_SAM_3
    #=GF ID   Phage_int_SAM_4
    #=GF ID   Phage_int_SAM_5
    #=GF ID   Phage_int_SAM_6
    #=GF ID   Phage_lambda_P
    #=GF ID   DUF3701
    #=GF ID   Phage_integrase
    #=GF ID   Phage_Integr_2
    #=GF ID   Phage_integr_3

----


 
 
### CDD batch search

**Input fasta ids (> lines) have to not have anything after a space. And the pipe symbol messes it up too.**  
Batch can handle 4000 queries.

CDD batch search with SMART does not detect the same integrase subfamilies - seems like it hasn't been updated since these new domains were added.

CDD domains match SMART domains?? Sort of but different names - not sure how they correspond yet.

*SMART has a batch annotation tool, though: 'You will only get results for sequences that have a match in SMART database, ie. only precomputed results will be displayed'.*


CDD has web api or can use stand-alone RPS-blast with CDD data from ftp.ncbi.nih.gov > cdd.  

HTTP GET command to send request
```{r, eval=FALSE}
library(httr)
base <- "https://www.ncbi.nlm.nih.gov/Structure/bwrpsb/bwrpsb.cgi?"

# databases: "cdd," "pfam," "smart," "tigrfam," "cog," "kog"
db <- 'db=CDD' 

# tdata: hits, aligns, or feats
tdata <- 'tdata=hits'

# data output: rep, std, or full
dmode <- 'dmode=std'

# sequences
# queries <- input_fasta
# 
# 
# HAVENT figure this API out yet...
```
 
 
 









### Other Figures


#### HMM model statistics

The hmm model statistics for each subfamily were recorded from the output of `hmmbuild` 
```{r hmm graphs, fig.width=8}
hmms_output <- read_csv('./data/SMART/hmm_output.csv')
glimpse(hmms_output)
# nseqs
a <- hmms_output |> 
  ggplot(aes(nseq, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n sequences',
       subtitle = 'Number of sequences for profile HMMs by sub-family')
# alignment columns
b <- hmms_output |>
  ggplot(aes(alen, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n columns',
       subtitle = 'Number of alignment columns by sub-family')
# number of consensus positions in hmm
c <- hmms_output |>
  ggplot(aes(mlen, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n nodes',
       subtitle = 'Number of positions in HMM by subfamily')
# effective number of sequences
d <- hmms_output |>
  ggplot(aes(eff_nseq, fct_rev(name))) +
  geom_col() +
  labs(y = '', x = 'n sequences',
       subtitle = 'Effective number of sequences by sub-family')

library(patchwork)
((a+b)|(c+d)) + plot_layout(nrow = 4)
```
 

 <!-- Hmmbuild output shows some data that might be of interest-->
 <!-- - n_eff: effective total seq number -->
 <!-- - re/pos: relative entropy per position -->
 <!-- .... -->
 








